inc/Engine/CriticalPath/APIClient.php

```diff
@@ -0,0 +1,230 @@
+					// translators: %s = item URL.
+						__( 'Critical CSS for %1$s not generated. Error: The API returned an empty response.', 'rocket' ),
+						$url
+					);
+				}
+				break;
+			case 400:
+			case 440:
+			case 404:
+				// translators: %s = item URL.
+				$message .= sprintf( __( 'Critical CSS for %1$s not generated.', 'rocket' ), $url );
+				break;
+			default:
+				$message .= sprintf(
+				// translators: %s = URL.
+					__( 'Critical CSS for %1$s not generated. Error: The API returned an invalid response code.', 'rocket' ),
+					$url
+				);
+				break;
+		}
+
+		if ( isset( $response_data->message ) ) {
+			// translators: %1$s = error message.
+			$message .= ' ' . sprintf( __( 'Error: %1$s', 'rocket' ), $response_data->message );
+		}
+
+		return $message;
+	}
+
+	/**
+	 * Get response data from the API.
+	 *
+	 * @since 3.6
+	 *
+	 * @param array|WP_Error $response The response or WP_Error on failure.
+	 * @return mixed response of API.
+	 */
+	private function get_response_data( $response ) {
+		return json_decode( wp_remote_retrieve_body( $response ) );
+	}
+
+	/**
+	 * Get our internal response code [Not the standard HTTP codes].
+	 *
+	 * @since 3.6
+	 *
+	 * @param array|WP_Error $response The response or WP_Error on failure.
+	 * @return string response code.
+	 */
+	private function get_response_code( $response ) {
+		// Todo: we can return code based on the response status number, for example 404 not_found.
+		return 'cpcss_generation_failed';
+	}
+
+	/**
+	 * Get job details by calling API with job ID.
+	 *
+	 * @since 3.6
+	 *
+	 * @param string $job_id ID for the job to get details.
+	 * @param string $url URL to be used in error messages.
+	 * @return mixed|WP_Error Details for job.
+	 */
+	public function get_job_details( $job_id, $url ) {
+		$response = wp_remote_get(
+			self::API_URL . "{$job_id}/"
+		);
+
+		return $this->prepare_job_details_response( $response, $url );
+	}
+
+	/**
+	 * Prepares Job details response to be returned.
+	 *
+	 * @since 3.6
+	 *
+	 * @param array|WP_Error $response The response or WP_Error on failure.
+	 * @param string         $url URL to be used in error messages.
+	 * @return mixed|WP_Error
+	 */
+	private function prepare_job_details_response( $response, $url ) {
+		$response_data        = $this->get_response_data( $response );
+		$response_status_code = $this->get_response_status( $response, ( isset( $response_data->status ) ) ? $response_data->status : null );
+		$succeeded            = $this->get_response_success( $response_status_code, $response_data );
+
+		if ( $succeeded ) {
+			return $response_data;
+		}
+
+		$response_code    = $this->get_response_code( $response );
+		$response_message = $this->get_response_message( $response_status_code, $response_data, $url );
+
+		if ( 200 === $response_status_code ) {
+			$response_status_code = 400;
+		}
+
+		return new WP_Error(
+			$response_code,
+			$response_message,
+			[
+				'status' => $response_status_code,
+			]
+		);
```

Comment:
This code is exactly the same as the code in `prepare_response()` method above. Refactor into a private method and then invoke it in both methods. Why? DRY.

Reply:
I merged two methods into one.